### ISSUE TITLE
scrcpy exits spontaneously a few times a day

### ISSUE THREAD

**Issue by freebrowser1** [Genymobile/scrcpy#4508]
- [ ] I have read the [FAQ](https://github.com/Genymobile/scrcpy/blob/master/FAQ.md).
 - [ ] I have searched in existing [issues](https://github.com/Genymobile/scrcpy/issues).

**Environment**
 - OS: macOS
 - scrcpy version: 2.2
 - installation method: from homebrew
 - device model: OnePlus 6
 - Android version: 10

**Describe the bug**
When connected wired via USB cable, it just exits a few times during a work day. After restart it runs a couple of hours and then it exits again spontaneously.
This does not happen when connected wirelessly.
```
% scrcpy -S -s <deviceID> -m 1024 -K
scrcpy v2.2 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  <deviceID>                        device  ONEPLUS_A6003
/Users/user/homebrew/Cellar/scrcpy/2.2/share/scrcpy/scrcpy-server: 1 file pushed, 0 skipped. 56.6 MB/s (64363 bytes in 0.001s)
[server] INFO: Device: [OnePlus] OnePlus ONEPLUS A6003 (Android 10)
[server] WARN: Audio disabled: it is not supported before Android 11
INFO: USB device found:
INFO:         0000000000000001   (2109:0100)  VIA Technologies Inc.          USB 2.0 BILLBOARD             
INFO:     --> <deviceID>           (2a70:9011)  OnePlus OnePlus
INFO:         058F84688461       (058f:8468)  Generic Mass Storage Device
INFO: USB device: <deviceID> (2a70:9011) OnePlus OnePlus
INFO: Renderer: metal
WARN: Demuxer 'audio': stream explicitly disabled by the device
INFO: Texture: 1024x504
[server] INFO: Device screen turned off
```
This terminal output appears when it spontaneously exits:
```
WARN: Device disconnected
ERROR: UNREGISTER_HID: libusb error: Input/Output Error
WARN: Could not unregister HID keyboard
```

**Comment by rom1v** [comment]
> WARN: Device disconnected

This happens when the adb connection gets closed.

This may happen due to [USB cable/port issue](https://github.com/Genymobile/scrcpy/blob/master/FAQ.md#device-disconnected), or when USB is reconfigured (when you enable/disable MTP for example).

This is not specific to scrcpy: if you run `adb shell` or `adb logcat` in a separate terminal, you'll observe they will terminate at the same time as scrcpy.

**Comment by rom1v** [comment]
> But adb should not terminate spontaneously.

From the adb point of vue, this is exactly as if the device was disconnected.

If you run an `adb` command and unplug the device, the adb command terminates.

> I never use MTP and never reconfigure USB when the device is connected using scrcpy.

Try with another cable or USB port then.

**Comment by freebrowser1** [comment]
The issue also occurs when connected via wifi, but less frequently.

**Comment by rom1v** [comment]
> The issue also occurs when connected via wifi, but less frequently.

Yes, this can happen too. Sometimes the adb server (in the device ROM) just crashes on the device (check your adb logcat).

Anyway, this is a problem at the adb level, if the adb connection is closed, scrcpy detects that the device is disconnected.

In theory, it could try to reconnect during a certain period, but it's too complicated for a workaround. This can be done by an external script which restarts if scrcpy terminates with an error (cf #721).

**Comment by shodanx2** [comment]
This solution might work for you
https://github.com/Genymobile/scrcpy/issues/721#issuecomment-2143601259
